lib/Mir.ml

```diff
@@ -145,18 +145,23 @@ type ('e, 's) prog =
   ; prog_path: string }
 [@@deriving sexp]
 
-type expr_typed_located =
-  { texpr_type: unsizedtype
-  ; texpr_loc: location_span sexp_opaque [@compare.ignore]
-  ; texpr: expr_typed_located expr
-  ; texpr_adlevel: autodifftype }
-[@@deriving sexp, hash, map, of_sexp]
-
-type stmt_loc =
-  { sloc: location_span sexp_opaque [@compare.ignore]
-  ; stmt: (expr_typed_located, stmt_loc) statement }
-[@@deriving hash, map, sexp, of_sexp]
+type 'm expr_with = {expr: 'm expr_with expr; emeta: 'm}
 
+and typed_and_located =
+  { expr_type: unsizedtype
+  ; expr_loc: location_span sexp_opaque [@compare.ignore]
+  ; expr_adlevel: autodifftype }
+
+and expr_typed_located = typed_and_located expr_with [@@deriving sexp]
```

Comment:
Do we need this alias? What if we renamed `typed_and_located` to `type_and_location`, and then write something like `type_and_location expr_with` in signatures, so it's more explicit what it is? (It's unfortunate that it doesn't read `expr_with type_and_location`, maybe there's a better phrasing, like renaming `expr_with` to `exprs` so we can write `type_and_location exprs`)

Reply:
type_and_location with_expr? :P I haven't really programmed with this pattern at a large scale before, it may turn out we don't need the alias. Usually would help make refactoring easier; for example I just changed what stmt_loc referred to and didn't have to change much code using that.